### PR TITLE
193

### PR DESCRIPTION
**Pull request by GregSutcliffe**
Ruby 1.9.3 fixes, brought up to date for latest develop

### CODE REVIEW

**Comment by ohadlevy** [comment]
@witlessb I just ran safemode tests and a few of them warn (only on 1.9), would you mind having a look? http://pastie.org/5501563

**Comment by GregSutcliffe** [comment]
Just updated to latest HEAD.

**Comment by ohadlevy** [comment]
I'm having a strange issue with bundler

when removing the lock file, bundle install wont work:

```
 bundle install
Updating https://github.com/jystewart/has_many_polymorphs.git
Fetching source index for http://rubygems.org/
Bundler could not find compatible versions for gem "ruby_parser":
  In Gemfile:
    safemode (~> 1.1.0) ruby depends on
      ruby_parser (~> 3.0.0) ruby

    safemode (~> 1.1.0) ruby depends on
      ruby_parser (3.1.0)
```

@domcleal suggested this is because we have a gemfile.lock in safemode repo 
https://github.com/svenfuchs/safemode/blob/master/Gemfile.lock#L20

Another issue with this patch is that tests can run! am I missing something?
/cc @witlessb

**Comment by domcleal** [comment]
I don't think it's the .lock causing the issue, but it points to the `~> 3.0.0` dependency coming from ruby2ruby.  I can't see where 3.1.0 is coming from though, except that it's the latest available.  3.0.4 would satisfy the dependency...

**Comment by GregSutcliffe** [comment]
I think rubygems was having some issues there - I can't replicate this, and apparently @domcleal can't either now...

**Comment by domcleal** [comment]
I can reproduce this again by removing both safemode and ruby_parser gems, then doing a `bundle install`.  However, with ruby_parser in our Gemfile with the existing `>= 2.3.1` specification seems to avoid this and bundler correctly installs safemode 1.1.0 and ruby_parser 3.0.4.

If we simply remove the unless conditional around ruby_parser, I think this will go away.

**Comment by GregSutcliffe** [comment]
@domcleal I still can't reproduce, but updated anyway :)

File: app/models/lookup_key.rb [hunk 1]
```diff
@@ -113,6 +113,14 @@ def as_json(options={})
     super({:only => [:key, :is_param, :required, :override, :description, :default_value, :id]}.merge(options))
   end
 
+  def path_elements
```
Comment (ohadlevy) [inline app/models/lookup_key.rb:116]: why this move?

**Comment by ohadlevy** [comment]
@GregSutcliffe can you also review & patch http://pastie.org/5515315

File: app/models/lookup_key.rb (hunk 1 above)
Comment (domcleal) [inline app/models/lookup_key.rb:116]: It's used in an Array#map call from the classification model.  It seems 1.8 allows you to call a private method from map (using the &: syntax) but 1.9 only lets you call public methods, hence it's moved above the private line.

File: bundler.d/test.rb [hunk 2]
```diff
@@ -1,8 +1,13 @@
...
+    gem 'mocha', :require => false
+  else
+    gem 'mocha', '< 0.13.0', :require => false
+  end
   gem 'shoulda', "=3.0.1"
   gem 'rr'
   gem 'rake'
   gem 'single_test'
   gem 'ci_reporter', '>= 1.6.3'
+  gem "minitest", '3.5.0', :platforms => :ruby_19
```
Comment (GregSutcliffe) [inline bundler.d/test.rb:12]: @ohadlevy what's this pin-version for?

**Comment by GregSutcliffe** [comment]
@ohadlevy pastie merged, looks mostly fine, one small question (inline)

File: bundler.d/test.rb (hunk 2 above)
Comment (ohadlevy) [inline bundler.d/test.rb:12]: I was unable to use any other version (due to conflict between mocha and
minitest). I took the first match from google/stackoverflow.

Comment (domcleal) [inline bundler.d/test.rb:12]: I ran tests ok earlier under 1.9 without the pinning, using mocha 0.13.1 and minitest 4.3.3.

Comment (ohadlevy) [inline bundler.d/test.rb:12]: OK, the version I used broke with the following error

```
rake
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`block in require': iconv will be deprecated in the future, use
String#encode instead.
/home/olevy/.gem/ruby/1.9.1/gems/puppet-2.7.19/lib/puppet/external/pson/pure.rb:7:in
`<module:PSON>': iconv will be deprecated in the future, use String#encode
instead.
/home/olevy/.gem/ruby/1.9.1/gems/mocha-0.12.7/lib/mocha/integration/mini_test.rb:56:in
`<class:TestCase>': No Mocha monkey-patch for MiniTest version
(RuntimeError)
from
/home/olevy/.gem/ruby/1.9.1/gems/mocha-0.12.7/lib/mocha/integration/mini_test.rb:20:in
`<class:Unit>'
from
/home/olevy/.gem/ruby/1.9.1/gems/mocha-0.12.7/lib/mocha/integration/mini_test.rb:19:in
`<module:MiniTest>'
from
/home/olevy/.gem/ruby/1.9.1/gems/mocha-0.12.7/lib/mocha/integration/mini_test.rb:18:in
`<top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`block in require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`block in load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:597:in
`new_constants_in'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/mocha-0.12.7/lib/mocha/integration.rb:45:in
`block in <top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/mocha-0.12.7/lib/mocha/integration.rb:44:in
`each'
from
/home/olevy/.gem/ruby/1.9.1/gems/mocha-0.12.7/lib/mocha/integration.rb:44:in
`<top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`block in require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`block in load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:597:in
`new_constants_in'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from /home/olevy/.gem/ruby/1.9.1/gems/mocha-0.12.7/lib/mocha.rb:2:in `<top
(required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`block in require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`block in load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:597:in
`new_constants_in'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/test_case.rb:11:in
`block in <top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/core_ext/kernel/reporting.rb:11:in
`block in silence_warnings'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/core_ext/kernel/reporting.rb:22:in
`with_warnings'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/core_ext/kernel/reporting.rb:11:in
`silence_warnings'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/test_case.rb:11:in
`<top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/actionpack-3.0.17/lib/action_controller/test_case.rb:310:in
`<module:ActionController>'
from
/home/olevy/.gem/ruby/1.9.1/gems/actionpack-3.0.17/lib/action_controller/test_case.rb:5:in
`<top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/shoulda-matchers-1.0.0/lib/shoulda/matchers/integrations/test_unit.rb:6:in
`<top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`block in require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`block in load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:597:in
`new_constants_in'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/shoulda-matchers-1.0.0/lib/shoulda/matchers.rb:7:in
`<top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`block in require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`block in load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:597:in
`new_constants_in'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/shoulda-matchers-1.0.0/lib/shoulda-matchers.rb:1:in
`<top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`block in require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`block in load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:597:in
`new_constants_in'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from /home/olevy/.gem/ruby/1.9.1/gems/shoulda-3.0.1/lib/shoulda.rb:2:in
`<top (required)>'
from /usr/share/gems/gems/bundler-1.0.21/lib/bundler/runtime.rb:68:in
`require'
from /usr/share/gems/gems/bundler-1.0.21/lib/bundler/runtime.rb:68:in
`block (2 levels) in require'
from /usr/share/gems/gems/bundler-1.0.21/lib/bundler/runtime.rb:66:in `each'
from /usr/share/gems/gems/bundler-1.0.21/lib/bundler/runtime.rb:66:in
`block in require'
from /usr/share/gems/gems/bundler-1.0.21/lib/bundler/runtime.rb:55:in `each'
from /usr/share/gems/gems/bundler-1.0.21/lib/bundler/runtime.rb:55:in
`require'
from /usr/share/gems/gems/bundler-1.0.21/lib/bundler.rb:122:in `require'
from /home/olevy/git/foreman/config/application.rb:11:in `<top (required)>'
from /home/olevy/git/foreman/config/environment.rb:2:in `require'
from /home/olevy/git/foreman/config/environment.rb:2:in `<top (required)>'
from /home/olevy/git/foreman/test/test_helper.rb:2:in `require'
from /home/olevy/git/foreman/test/test_helper.rb:2:in `<top (required)>'
from /home/olevy/git/foreman/test/unit/architecture_test.rb:1:in `require'
from /home/olevy/git/foreman/test/unit/architecture_test.rb:1:in `<top
(required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/rake-10.0.2/lib/rake/rake_test_loader.rb:10:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/rake-10.0.2/lib/rake/rake_test_loader.rb:10:in
`block (2 levels) in <main>'
from
/home/olevy/.gem/ruby/1.9.1/gems/rake-10.0.2/lib/rake/rake_test_loader.rb:9:in
`each'
from
/home/olevy/.gem/ruby/1.9.1/gems/rake-10.0.2/lib/rake/rake_test_loader.rb:9:in
`block in <main>'
from
/home/olevy/.gem/ruby/1.9.1/gems/rake-10.0.2/lib/rake/rake_test_loader.rb:4:in
`select'
from
/home/olevy/.gem/ruby/1.9.1/gems/rake-10.0.2/lib/rake/rake_test_loader.rb:4:in
`<main>'
/home/olevy/.gem/ruby/1.9.1/gems/mocha-0.12.7/lib/mocha/integration/mini_test.rb:56:in
`<class:TestCase>': No Mocha monkey-patch for MiniTest version
(RuntimeError)
from
/home/olevy/.gem/ruby/1.9.1/gems/mocha-0.12.7/lib/mocha/integration/mini_test.rb:20:in
`<class:Unit>'
from
/home/olevy/.gem/ruby/1.9.1/gems/mocha-0.12.7/lib/mocha/integration/mini_test.rb:19:in
`<module:MiniTest>'
from
/home/olevy/.gem/ruby/1.9.1/gems/mocha-0.12.7/lib/mocha/integration/mini_test.rb:18:in
`<top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`block in require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`block in load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:597:in
`new_constants_in'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/mocha-0.12.7/lib/mocha/integration.rb:45:in
`block in <top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/mocha-0.12.7/lib/mocha/integration.rb:44:in
`each'
from
/home/olevy/.gem/ruby/1.9.1/gems/mocha-0.12.7/lib/mocha/integration.rb:44:in
`<top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`block in require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`block in load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:597:in
`new_constants_in'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from /home/olevy/.gem/ruby/1.9.1/gems/mocha-0.12.7/lib/mocha.rb:2:in `<top
(required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`block in require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`block in load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:597:in
`new_constants_in'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/test_case.rb:11:in
`block in <top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/core_ext/kernel/reporting.rb:11:in
`block in silence_warnings'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/core_ext/kernel/reporting.rb:22:in
`with_warnings'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/core_ext/kernel/reporting.rb:11:in
`silence_warnings'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/test_case.rb:11:in
`<top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/actionpack-3.0.17/lib/action_controller/test_case.rb:310:in
`<module:ActionController>'
from
/home/olevy/.gem/ruby/1.9.1/gems/actionpack-3.0.17/lib/action_controller/test_case.rb:5:in
`<top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/shoulda-matchers-1.0.0/lib/shoulda/matchers/integrations/test_unit.rb:6:in
`<top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`block in require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`block in load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:597:in
`new_constants_in'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/shoulda-matchers-1.0.0/lib/shoulda/matchers.rb:7:in
`<top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`block in require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`block in load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:597:in
`new_constants_in'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/shoulda-matchers-1.0.0/lib/shoulda-matchers.rb:1:in
`<top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`block in require'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`block in load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:597:in
`new_constants_in'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:225:in
`load_dependency'
from
/home/olevy/.gem/ruby/1.9.1/gems/activesupport-3.0.17/lib/active_support/dependencies.rb:242:in
`require'
from /home/olevy/.gem/ruby/1.9.1/gems/shoulda-3.0.1/lib/shoulda.rb:2:in
`<top (required)>'
from /usr/share/gems/gems/bundler-1.0.21/lib/bundler/runtime.rb:68:in
`require'
from /usr/share/gems/gems/bundler-1.0.21/lib/bundler/runtime.rb:68:in
`block (2 levels) in require'
from /usr/share/gems/gems/bundler-1.0.21/lib/bundler/runtime.rb:66:in `each'
from /usr/share/gems/gems/bundler-1.0.21/lib/bundler/runtime.rb:66:in
`block in require'
from /usr/share/gems/gems/bundler-1.0.21/lib/bundler/runtime.rb:55:in `each'
from /usr/share/gems/gems/bundler-1.0.21/lib/bundler/runtime.rb:55:in
`require'
from /usr/share/gems/gems/bundler-1.0.21/lib/bundler.rb:122:in `require'
from /home/olevy/git/foreman/config/application.rb:11:in `<top (required)>'
from /home/olevy/git/foreman/config/environment.rb:2:in `require'
from /home/olevy/git/foreman/config/environment.rb:2:in `<top (required)>'
from /home/olevy/git/foreman/test/test_helper.rb:2:in `require'
from /home/olevy/git/foreman/test/test_helper.rb:2:in `<top (required)>'
from
/home/olevy/git/foreman/test/functional/api/v1/architectures_controller_test.rb:1:in
`require'
from
/home/olevy/git/foreman/test/functional/api/v1/architectures_controller_test.rb:1:in
`<top (required)>'
from
/home/olevy/.gem/ruby/1.9.1/gems/rake-10.0.2/lib/rake/rake_test_loader.rb:10:in
`require'
from
/home/olevy/.gem/ruby/1.9.1/gems/rake-10.0.2/lib/rake/rake_test_loader.rb:10:in
`block (2 levels) in <main>'
from
/home/olevy/.gem/ruby/1.9.1/gems/rake-10.0.2/lib/rake/rake_test_loader.rb:9:in
`each'
from
/home/olevy/.gem/ruby/1.9.1/gems/rake-10.0.2/lib/rake/rake_test_loader.rb:9:in
`block in <main>'
from
/home/olevy/.gem/ruby/1.9.1/gems/rake-10.0.2/lib/rake/rake_test_loader.rb:4:in
`select'
from
/home/olevy/.gem/ruby/1.9.1/gems/rake-10.0.2/lib/rake/rake_test_loader.rb:4:in
`<main>'
Errors running test:units, test:functionals!
```

Comment (domcleal) [inline bundler.d/test.rb:12]: mocha 0.12.7 looks old, perhaps try an update to 0.13.x?

Comment (domcleal) [inline bundler.d/test.rb:12]: https://github.com/freerange/mocha/blob/master/RELEASE.md indicates 0.13.x doesn't need monkey-patching any more, but doesn't indicate which minimum version of minitest is required.  We could probably change mocha to >= 0.13.0 then, if that works for ruby 1.9 with a "recent" version of minitest.

Comment (GregSutcliffe) [inline bundler.d/test.rb:12]: I don't have a real issue with it, I was just curious about the specific version pin - normally we'd use '~' or '>='

**Comment by GregSutcliffe** [comment]
Just confirmed that my 1.9 foreman host seems fine with the current state of this branch.

**Comment by aboe76** [comment]
fixed it with the following line: 
gem "safemode", ">=1.0.0","<=1.1.0"

and got this installed:

# bundle list

Gems included by the bundle:
- abstract (1.0.0)
- actionmailer (3.0.17)
- actionpack (3.0.17)
- activemodel (3.0.17)
- activerecord (3.0.17)
- activeresource (3.0.17)
- activesupport (3.0.17)
- ancestry (1.3.0)
- apipie-rails (0.0.13)
- archive-tar-minitar (0.5.2)
- arel (2.0.10)
- audited (3.0.0.rc1)
- audited-activerecord (3.0.0.rc1)
- awesome_print (1.1.0)
- builder (2.1.2)
- bundler (1.2.2)
- ci_reporter (1.8.3)
- coderay (1.0.8)
- columnize (0.3.6)
- erubis (2.6.6)
- excon (0.16.10)
- fog (1.8.0)
- formatador (0.2.4)
- has_many_polymorphs (3.0.0.beta1 03429a6)
- hirb (0.7.0)
- hirb-unicode (0.0.5)
- i18n (0.5.0)
- jquery-rails (2.1.4)
- json (1.7.5)
- linecache19 (0.5.12)
- mail (2.2.19)
- metaclass (0.0.1)
- method_source (0.8.1)
- mime-types (1.19)
- minitest (3.5.0)
- mocha (0.13.1)
- multi_json (1.5.0)
- mysql (2.9.0)
- mysql2 (0.2.18)
- net-ldap (0.3.1)
- net-scp (1.0.4)
- net-ssh (2.6.2)
- nokogiri (1.5.5)
- oauth (0.4.7)
- paint (0.8.5)
- pg (0.14.1)
- polyglot (0.3.3)
- pry (0.9.10)
- rabl (0.7.9)
- rack (1.2.5)
- rack-mini-profiler (0.1.23)
- rack-mount (0.6.14)
- rack-test (0.5.7)
- rails (3.0.17)
- railties (3.0.17)
- rake (10.0.2)
- rbovirt (0.0.16)
- rbvmomi (1.6.0)
- rdoc (3.12)
- redcarpet (2.1.0)
- rest-client (1.6.7)
- rr (1.0.4)
- ruby-debug-base19 (0.11.25)
- ruby-debug19 (0.11.6)
- ruby-hmac (0.4.0)
- ruby-libvirt (0.4.0)
- ruby2ruby (2.0.1)
- ruby_core_source (0.1.5)
- ruby_parser (3.0.4)
- safemode (1.1.0)             <<<< still 1.1.0 without the depency error.
- scoped_search (2.4.0)
- sexp_processor (4.1.3)
- shoulda (3.0.1)
- shoulda-context (1.0.1)
- shoulda-matchers (1.0.0)
- single_test (0.5.1)
- slop (3.3.3)
- sqlite3 (1.3.6)
- term-ansicolor (1.0.7)
- thor (0.14.6)
- treetop (1.4.12)
- trollop (2.0)
- tzinfo (0.3.35)
- unicode-display_width (0.1.1)
- uuidtools (2.1.3)
- will_paginate (3.0.3)
- wirb (1.0.1)

**Comment by ohadlevy** [comment]
isnt that the same like saying

gem "safemode", "1.1.0" # since there were only two releases 1.0.0 and 1.1.0 ?

**Comment by aboe76** [comment]
I know, but was searching for the error on google, and found this answer and tried it...
lo and behold it worked...what kind of black magic this is I don't know.

**Comment by ohadlevy** [comment]
ok, so there are two different issues here, 1. is with the Safemode bundle and the other with minitest, @GregSutcliffe @domcleal what would you recommend then?

**Comment by GregSutcliffe** [comment]
@ohadlevy minitest is just for the testing bundle, so I think we can roll-forward if we find problems with it. Safemode is more interesting - I still cannot replicate the problem, but I can add the fix suggested by @aboe76 if we think it's acceptable. @domcleal ?

**Comment by GregSutcliffe** [comment]
forgot to mention: i think this needs to go into rc2 so we have time to find issues with it before final release. I don't want to see ruby1.9 support pushed back to 1.2 :)

**Comment by ohadlevy** [comment]
@domcleal @aboe76 sadly this does not solve the problem for me, it either works for ruby 1.9 or 1.87 not both.

sadly i cant merge it as is, and maybe we need to relax safemode requirement

**Comment by domcleal** [comment]
Been playing with this and I'm able to install under both 1.8.7 and 1.9.3 if I specify:

```
gem "safemode", "~> 1.1.0"
gem 'ruby_parser', '~> 3.0.0'
```

(i.e. sort of duplicating the requirement from safemode)  I still can't see where the 3.1.0 requirement is coming from, since the rubygems API is reporting it correctly - a bug in the bundler resolver perhaps.

How does that work for you @ohadlevy and @aboe76?

**Comment by ohadlevy** [comment]
@domcleal sadly it does not work on ruby 1.93 - i still get

```
Bundler could not find compatible versions for gem "ruby_parser":
  In Gemfile:
    safemode (~> 1.1.0) ruby depends on
      ruby_parser (~> 3.0.0) ruby

    safemode (~> 1.1.0) ruby depends on
      ruby_parser (3.1.0)


```

**Comment by domcleal** [comment]
@ohadlevy hm, I can't reproduce this on a clean 1.9.3 gemset.  Sigh.

**Comment by ohadlevy** [comment]
@domcleal which version of bundler do you use?

**Comment by domcleal** [comment]
@ohadlevy 1.2.3

**Comment by ohadlevy** [comment]
OK, upgrading bundler to 1.2.3  fixed it for me!

@skottler @GregSutcliffe what do you guys think? can we enforce bundle version too?

**Comment by GregSutcliffe** [comment]
I'm already packaging Bundler 1.2.2 for squeeze/precise (http://deb.theforeman.org/pool/nightly/b/bundler/) - it's needed for the vendored gems packages I have in nightly at the moment. I'm happy to depend on that - @skottler ?

**Comment by skottler** [comment]
We current have bundler 1.0.15 in the RC repo, but I'm happy to update that
when we release the 2nd RC.

**Comment by ohadlevy** [comment]
rebased and merged 
thanks @witlessbird @skottler @GregSutcliffe @domcleal @aboe76

**Pull request closed without merging**